### ISSUE TITLE
Async client can't publish at a high rate (20Hz) after v1.3.5 

### ISSUE THREAD

**Issue by beville** [eclipse/paho.mqtt.c#1252]
**Summary:** 

Using paho-mqtt-c (1.3.10) / paho-mqtt-cpp (1.2.0), publishing messages at a fast rate (about 20/sec) using async client can't keep up after upgrading paho-mqtt-c from 1.3.5 to a higher version.

**Details:**

We have Linux (5.10) devices that are using Yocto builds that have older versions of the paho libraries (paho-mqtt-c:1.3.1) / paho-mqtt-cpp :1.0.0), and our application code is able to publish at 20 messages a second by by setting the the maxInflightMessages to 10.  The actual inflight count never gets above 5 in practice, and there is never a problem keeping up.  We're not using the persistence features of paho-mqtt, but have a queue that we use before paho publish calls, and don't publish more than the maxInflightMessages by keeping track of delivery token ids from the `publish` method and `delivery_complete` callbacks.

However, once we upgraded our paho packages, we saw that the publish rate with the same application code couldn't be sustained.  The effective rate was more like 10/sec, and once publishing stopped, the remaining 10 in-flight messages were published at 1/sec.

We were able to narrow down the version where we see this issue start was paho-mqtt-c 1.3.6.  So:

- paho-mqtt-c 1.3.5 / paho-mqtt-cpp 1.0.0:  Able to publish at 20hz
- paho-mqtt-c 1.3.6 / paho-mqtt-cpp 1.0.0: **_Not_** able to publish at 20hz
- paho-mqtt-c 1.3.10 / paho-mqtt-cpp 1.2.0: **_Not_** able to publish at 20hz

 Before I try to create a clean example to demonstrate this (since I'll have re-create it from our production code, which will be a bit of work), I want to first verify:

- Do you consider this a bug?  Maybe that kind of publish rate isn't supported?
- Do you know why the behavior might have changed between 1.3.5 and 1.36?  I've looked at the diffs between versions, and nothing jumped out at me. 
- Since we're using the CPP wrapper classes, should I post this issue at (https://github.com/eclipse/paho.mqtt.cpp/) instead?

**Comment by icraggs** [comment]
That frequency sounds very low to me but of course it depends on your hardware, QoS, whether you're using TLS or not, where the MQTT server is, and so on.

A recent test I wrote using QoS 2, no TLS, local Mosquitto broker, HICH_PERFORMANCE=TRUE, waiting for message to be returned from a subscription before sending the next one, gets about 2500 messages per second on a pretty powerful desktop.

There is a timing issue in the 1.3.10 release that causes some scenarios to be rate limited in the new poll implementation. I've fixed this in the develop branch. Alternatively, if you build 1.3.10 with the CMake option USE_SELECT=TRUE, this should perform similarly.

**Comment by beville** [comment]
Thanks for the quick response on this!

Sorry, I should have provided more context: These are Arm devices communicating to remote brokers via cell modems using TLS. The cloud-based broker handles maybe 100k connections from the devices at any moment. From what I observed with this setup, the latency (when sending one message at a time) between a QoS 1 PUBLISH and a PUBACK is over 100ms, often around 140ms.

```
20220725 093110.028 17 mqttbridge-QM00000155 -> PUBLISH msgid: 3 qos: 1 retained: 0 rc 0 payload len(168): �
20220725 093110.169 17 mqttbridge-QM00000155 <- PUBACK msgid: 3

20220725 093110.221 17 mqttbridge-QM00000155 -> PUBLISH msgid: 4 qos: 1 retained: 0 rc 0 payload len(46): *
20220725 093110.370 17 mqttbridge-QM00000155 <- PUBACK msgid: 4
```

I also failed to mention that I did one test where I had used the latest version of paho (1.3.10) to talk to a mosquitto broker over a LAN with no SSL, and there were no issues at all at 20Hz.

I just tried a build of 1.3.10 with USE_SELECT=TRUE in our production environment (with the remote broker and high latency PUBACKs) and it still couldn't keep up at 20Hz msgs with max in-flight set to 20.

As it turns out, I've been informed that this isn't a priority here, though, so I guess I can't devote more time to this, but I wanted to leave you with what I've observed.

Thanks again!
